src/services/navigator/helpers.ts

```diff
@@ -301,3 +302,132 @@ export const buildRequest = (
 
   return [navigation, lastPathId || routeId, params];
 };
+
+/**
+ * Create a map of <id, element> from a list of nodes
+ */
+const nodesToMap = (
+  nodes: TypesLegacy.NodeList<TypesLegacy.Node>,
+): Map<string, TypesLegacy.Node> => {
+  const map: Map<string, TypesLegacy.Node> = new Map();
+  if (!nodes) {
+    return map;
+  }
+  Array.from(nodes).forEach(node => {
+    if (node.nodeType === TypesLegacy.NODE_TYPE.ELEMENT_NODE) {
+      const element = node as TypesLegacy.Element;
+      if (isNavigationElement(element)) {
+        const id = element.getAttribute('id');
+        if (id) {
+          map.set(id, element);
+        }
+      }
+    }
+  });
+  return map;
+};
+
+const KEY_MERGE = 'merge';
+const KEY_INITIAL = 'initial';
+
+/**
+ * Merge the nodes from the new document into the current
+ * All attributes in the current are reset (initial, merge)
+ * If an id is found in both docs, the current node is updated
+ * If an id is found only in the new doc, the node is added to the current
+ * the 'merge' attribute on a navigator determines if the children are merged or replaced
+ */
+const mergeNodes = (
+  current: TypesLegacy.Element,
+  newNodes: TypesLegacy.NodeList<TypesLegacy.Node>,
+): void => {
+  if (!current || !current.childNodes || !newNodes || newNodes.length === 0) {
+    return;
+  }
+
+  // Clean out current node attributes for 'merge' and 'initial'
+  Array.from(current.childNodes).forEach(node => {
+    const element = node as TypesLegacy.Element;
+    if (isNavigationElement(element)) {
+      if (element.localName === TypesLegacy.LOCAL_NAME.NAVIGATOR) {
+        element.setAttribute(KEY_MERGE, 'false');
+      } else if (element.localName === TypesLegacy.LOCAL_NAME.NAV_ROUTE) {
+        element.setAttribute(KEY_INITIAL, 'false');
+      }
+    }
+  });
+
+  const currentMap = nodesToMap(current.childNodes);
+
+  Array.from(newNodes).forEach(node => {
+    if (node.nodeType === TypesLegacy.NODE_TYPE.ELEMENT_NODE) {
+      const newElement = node as TypesLegacy.Element;
+      if (isNavigationElement(newElement)) {
+        const id = newElement.getAttribute('id');
+        if (id) {
+          const currentElement = currentMap.get(id) as TypesLegacy.Element;
+          if (currentElement) {
+            if (newElement.localName === TypesLegacy.LOCAL_NAME.NAVIGATOR) {
+              const isMergeable =
+                newElement.getAttribute('merge')?.toLowerCase() === 'true';
```

Comment:
Minor, but I think accepted values for boolean attributes are `true`, `false`, `1`, `0`. We shouldn't need to do the case conversion. We don't explicitly support `1` and `0` anywhere I believe.

Reply:
Removed in a few places